### PR TITLE
Fix for hanging in D-Bus code

### PR DESCRIPTION
**Pull request by jbaublitz**

### CODE REVIEW

**Comment by jbaublitz** [comment]
@mulkieran This is not quite ready for review yet but I wanted to put up what I have so far. There's an ugly panic message that happens when `ctrl-C` is issued. I think I can resolve this by changing the `Arc<AtomicBool>` to a `tokio::sync::watch::channel`-based notification when `ctrl-C` is received. My understanding is that the panic message is due to the timeout future trying to start on a Runtime that's already shutting down. The channel-based approach should allow us to block without a timeout and select whichever future returns first. I'll give that a shot tomorrow.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks good so far. Two minor comments. Will need one more pass.

File: src/stratis/run.rs [hunk 1]
```diff
@@ -4,32 +4,37 @@
...
+    engine::{Lockable, LockableEngine, SimEngine, StratEngine, UdevEngineEvent},
     stratis::{
         dm::dm_event_thread, errors::StratisResult, ipc_support::setup, stratis::VERSION,
         udev_monitor::udev_thread,
     },
 };
 
 // Waits for SIGINT. If received, sets should_exit to true.
-async fn signal_thread(should_exit: Arc<AtomicBool>) {
+async fn signal_thread(trigger: Sender<bool>) {
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/stratis/run.rs:28]: Ah yes! Will do.

File: src/stratis/run.rs [hunk 2]
```diff
@@ -57,34 +62,33 @@ pub fn run(sim: bool) -> StratisResult<()> {
...
                         error!("Failed to start up stratisd engine: {}; exiting", e);
                         return Err(e);
                     }
-                }))
+                })
             }
         };
 
-        let should_exit = Arc::new(AtomicBool::new(false));
+        let (trigger, should_exit) = channel(2);
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/stratis/run.rs:82]: We only have two instances of the sender side of things and each one can only ever send exactly one message. I've verified that we will never send more than two for that reason. We can increase it, but it doesn't add anything from my point of view.

File: src/stratis/run.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/stratis/run.rs:29]: Resolved.

File: src/stratis/run.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/stratis/run.rs:83]: I actually think I missed adding in the second `send` invocation after the `select` macro in `run.rs`. I'll add that. Thanks for catching that!

**Review by mulkieran — CHANGES_REQUESTED** [review]
This looks ready to go! Please close and open a final PR.

**Review by mulkieran — CHANGES_REQUESTED** [review]
I ran into an odd problem; please see https://github.com/stratis-storage/stratisd/pull/2590 for a suggestion that might make sense.

The problem was an occasional ERROR log entry 

```[2021-05-12T20:01:39Z ERROR libstratis::stratis::ipc_support::dbus_support] The tree handling thread exited...```

on shutting down stratisd. The change I've put up seems to make that intermittent error disappear.

**Comment by mulkieran** [comment]
@jbaublitz Other than commit https://github.com/stratis-storage/stratisd/pull/2590/commits/6acc957a359fbb176fb37452298bf15b806365bc suggestion, this looks ready to go. Feel free to drop the ```debug!``` line, I don't think it's really all that useful.

**Comment by jbaublitz** [comment]
@mulkieran I changed the commit slightly because the thread exit logging has always been a little strange. I'll wait till you take a look at the most recent commit to put up a final PR.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks good! Please close and open final PR.

**Pull request closed without merging**